crates/ruff_benchmark/benches/red_knot.rs

```diff
@@ -24,40 +24,11 @@ const TOMLLIB_312_URL: &str = "https://raw.githubusercontent.com/python/cpython/
 
 static EXPECTED_DIAGNOSTICS: &[&str] = &[
     // We don't support `ModuleType`-attributes as globals yet:
-    "/src/tomllib/__init__.py:10:30: Name `__name__` used when not defined",
     // We don't support `*` imports yet:
     "/src/tomllib/_parser.py:7:29: Module `collections.abc` has no member `Iterable`",
     // We don't support terminal statements in control flow yet:
     "/src/tomllib/_parser.py:353:5: Method `__getitem__` of type `Unbound | @Todo` is not callable on object of type `Unbound | @Todo`",
     "/src/tomllib/_parser.py:455:9: Method `__getitem__` of type `Unbound | @Todo` is not callable on object of type `Unbound | @Todo`",
-    // True positives!
-    "Line 69 is too long (89 characters)",
-    "Use double quotes for strings",
-    "Use double quotes for strings",
-    "Use double quotes for strings",
-    "Use double quotes for strings",
-    "Use double quotes for strings",
-    "Use double quotes for strings",
-    "Use double quotes for strings",
-    // We don't support terminal statements in control flow yet:
-    "/src/tomllib/_parser.py:66:18: Name `s` used when possibly not defined",
-    "/src/tomllib/_parser.py:98:12: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:101:12: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:104:14: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:104:14: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:115:14: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:115:14: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:126:12: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:348:20: Name `nest` used when possibly not defined",
-    "/src/tomllib/_parser.py:353:5: Name `nest` used when possibly not defined",
-    "/src/tomllib/_parser.py:453:24: Name `nest` used when possibly not defined",
-    "/src/tomllib/_parser.py:455:9: Name `nest` used when possibly not defined",
-    "/src/tomllib/_parser.py:482:16: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:566:12: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:573:12: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:579:12: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:580:63: Name `char` used when possibly not defined",
-    "/src/tomllib/_parser.py:629:38: Name `datetime_obj` used when possibly not defined"
```

Comment:
I agree that the lint rules complaining about double quotes and lines being too long just add noise and distraction. But these lint rules seem like a classic case of "lints" which are core to a type checker's purpose: if we don't emit these lint diagnostics, we can't say for sure that the program is type-safe (and if we're not capable of doing this analysis significantly better than Ruff's current semantic model, we've failed).

Moreover, without these lint rules being emitted, I think we're losing valuable signal about the accuracy of our control-flow analysis

Reply:
Here, for example, these diagnostics indicate a serious deficiency in our control-flow analysis for exception handlers due to the fact that we don't yet understand terminal statements. That's why it's one of my monthly goals to work on that!

Reply:
I agree that this rule is useful and helps highlight deficiencies in our control flow analysis. However, I think it should be implemented as part of the type checker rather than as a separate linter phase.

Reply:
Okay, that makes sense.

I would prefer to reimplement the rule as part of the type checker in this PR: otherwise we'll have to just add these entries back to the "expected diagnostics" list in the followup PR when we reimplement the rule. Doing it as part of the same PR reduces churn in this file. It should also help us see whether the performance improvement currently being reported is due to removing the lint phase or due to removing this rule.

The rule is also very simple; it shouldn't be hard to reimplement it in this PR (just a matter of copying a couple of lines that we currently have in `crates/red_knot_workspace/src/lint.rs` to `crates/red_knot_python_semantic/src/types/infer.rs`).

On the other hand, if you _do_ get rid of the rule entirely in this PR, you also need to get rid of the comment on line 26 :-)

Reply:
I'm not very eager on implementing the rule because i worry that you'll ask me to add tests but I did so anyway

Reply:
Oh god, this breaks so many mdtests

Reply:
Okay, moving this into `infer` is more work than I currently plan spending on this because it breaks many mdtests that use undefined variables. 

and just doing in `infer.rs` is too naive

```
                let ty = bindings_ty(self.db, definitions, unbound_ty);

                if ty.is_unbound() {
                    self.add_diagnostic(
                        name.into(),
                        "unresolved-reference",
                        format_args!("Name `{}` used when not defined", id),
                    );
                } else if ty.may_be_unbound(self.db) {
                    self.add_diagnostic(
                        name.into(),
                        "unresolved-reference",
                        format_args!("Name `{}` used when possibly not defined", id),
                    );
                }

                ty
```

We can either close this PR, or create an issue to implement said rule in `infer` and do all the follow up work

Reply:
This illustrates to me that it's very important that we make this change soon! It makes sense to me that we should do this lint in the type-checking phase, and that errors from this lint should be reported in our mdtests. (If we _don't_ report this lint in our mdtests, it opens up the possibility of us not testing what we wanted to test, because a variable in an mdtest might be undefined when we didn't want it to be.)

So I think we'll have to update our mdtests to make them resilient to this rule at some point. And that means that we should make the change sooner rather than later (because the longer we defer it, the more mdtests we'll have to update when we do make the change).

If you're okay with me pushing to your branch, I'd be happy to do the busywork here of moving the rule to `infer.rs` and updating the mdtests.

Reply:
> If you're okay with me pushing to your branch, I'd be happy to do the busywork here of moving the rule to infer.rs and updating the mdtests.

Sure, feel free to go ahead (generally speaking, I don't mind other people pushing to my branch)

Reply:
I assume this is what you're already planning, @AlexWaygood , but I hope that we can in most cases actually update the mdtests to not use undefined variables, rather than updating them to expect the undefined variable error.

Reply:
> I assume this is what you're already planning, @AlexWaygood , but I hope that we can in most cases actually update the mdtests to not use undefined variables, rather than updating them to expect the undefined variable error.

Yes, of course!!